example/spec/helpers/apiUtils.js

```diff
@@ -0,0 +1,76 @@
+const fs = require('fs-extra');
```

Comment:
To give some context to this file, the scope of this ticket includes testing whether the URLs published to CMR are downloadable. In order to do that testing, I needed the URLs posted to CMR by the tests to be accessible from the tests. The easiest way to facilitate that was to run the local Express server for the distribution API at `http://localhost:5002` and setup the tests to publish distribution API URLs using that host. The relevant tests are IngestGranuleSuccessSpec.js and IngestUMMGSuccessSpec.js.

This worked great when running the tests individually, but when I tried to run them in parallel as Travis does, I got errors because whichever test was faster would initialize the express server and the second one would throw an `EADDRINUSE` error trying to listen on the same port. 

Ultimately, resolving the problem proved difficult given that each test is running in a separate Node process and as a separate run of Jasmine. I was able to create a node script for initializing the distribution API and kick it off from our `run-integration-tests.sh` shell script, but even though I could capture the PID of the process running the Express server, I didn't see a way to access that PID in a later Travis script to kill the process (e.g. `cleanup-integration-tests.sh`). This may not matter ultimately on Travis since I think the machine running the tests gets killed once it is finished and thus leftover processes don't matter, but it would matter for running the tests locally. And running the tests individually would have required some other work to coordinate starting up the Express server.

The only solution I came to is creating a helper to kick off the Express server in a child process. And then I store the PID of Express server process along with whatever tests are actively listening to it in a JSON file (allowing coordination across Node processes), so that I can know when the server is already running and when it is safe to shut it down.

The upside is that it seems to work and it allows tests to manage their own dependency on the distribution API server, plus it preserves the parallel test processing. The downside is that it's a fair amount of code, albeit simple, to accomplish those goals, so it may be unnecessary technical debt.

The other alternative @laurenfrederick suggested is just moving one of the tests, perhaps IngestUMMGSuccessSpec.js, to not process in parallel. That test is pretty fast (~60s), so it wouldn't add much to our build times. And then each test could start up/shut down the Express server separately. I'm happy to go with that solution if we want to keep things simpler (as I write this it seems like that might be better, all things considered). The only downside would be having to do the same for any future tests that need to use the distribution API..

Reply:
Yes, killing tests would mess with the approach, which is probably a point in favor.

As for using a separate script, I was able to create a node script for initializing the distribution API and kick it off from our run-integration-tests.sh shell script, but even though I could capture the PID of the process running the Express server, I didn't see a way to access that PID in a later Travis script to kill the process (e.g. cleanup-integration-tests.sh). This may not matter ultimately on Travis since I think the machine running the tests gets killed once it is finished and thus leftover processes don't matter, but it would matter for running the tests locally. And running the tests individually would have required some other work to coordinate starting up the Express server.

All that being said, maybe we axe this and move the faster tests that use the distribution API into serial?